### PR TITLE
modernize the cmake files a bit

### PR DESCRIPTION
**Pull request by ClausKlein**
try to use a namespace for cmake config package

prevent use of set_target_properties() if possible too

this fix https://github.com/TheLartians/ModernCppStarter/issues/79

because of https://github.com/TheLartians/ModernCppStarter/issues/78 there are format issues

### CODE REVIEW

**Comment by ClausKlein** [comment]
It is common to install cmake config packages with namespaces. i.e. `fmt::fmt`, and `doctets::doctest`

How can I handle this with **`CPMAddPackage()`**?

File: CMakeLists.txt [hunk 1]
```diff
@@ -30,27 +36,38 @@ CPMAddPackage(
...
-# INTERFACE_COMPILE_FEATURES cxx_std_17)
-
-add_library(Greeter ${headers} ${sources})
+# target:
+add_library(Greeter)
+add_library(${PROJECT_NAME}::Greeter ALIAS Greeter)
+target_compile_features(Greeter PUBLIC cxx_std_17)
 
-set_target_properties(Greeter PROPERTIES CXX_STANDARD 17)
+target_sources(Greeter PRIVATE ${headers} ${sources})
```
Comment (TheLartians) [inline CMakeLists.txt:64]: What is the advantage of `target_sources` vs adding them at the library definition?

File: CMakeLists.txt [hunk 2]
```diff
@@ -30,27 +36,38 @@ CPMAddPackage(
...
-file(GLOB_RECURSE headers CONFIGURE_DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/include/*.h")
-file(GLOB_RECURSE sources CONFIGURE_DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/source/*.cpp")
+file(GLOB_RECURSE headers CONFIGURE_DEPENDS "include/*.h")
+file(GLOB_RECURSE sources CONFIGURE_DEPENDS "source/*.cpp")
 
 # ---- Create library ----
 
 # Note: for header-only libraries change all PUBLIC flags to INTERFACE and create an interface
-# target: add_library(Greeter INTERFACE) set_target_properties(Greeter PROPERTIES
-# INTERFACE_COMPILE_FEATURES cxx_std_17)
```
Comment (TheLartians) [inline CMakeLists.txt:44]: This is supposed to be a template for header only libraries. Unfortunately, `cmake-format` has destroyed the formatting, so you have a point that we should reformat the comment.

File: CMakeLists.txt [hunk 3]
```diff
@@ -72,5 +89,8 @@ packageProject(
   INCLUDE_DIR ${PROJECT_SOURCE_DIR}/include
   INCLUDE_DESTINATION include/${PROJECT_NAME}-${PROJECT_VERSION}
   VERSION_HEADER "${VERSION_HEADER_LOCATION}"
-  DEPENDENCIES ""
+  # semicolon separated list of the project's dependencies
+  DEPENDENCIES fmt
+  # install your library with a namespace! (do NOT add extra '::')
+  NAMESPACE ${PROJECT_NAME}
```
Comment (TheLartians) [inline CMakeLists.txt:95]: The `Namespace` argument is already passed above

File: cmake/tools.cmake [hunk 4]
```diff
@@ -65,10 +67,11 @@ if(USE_SANITIZER OR USE_STATIC_ANALYZER)
 endif()
 
 # enables CCACHE support through the USE_CCACHE flag possible values are: YES, NO or equivalent
+option(USE_CCACHE "enable ccache usages" ON)
```
Comment (TheLartians) [inline cmake/tools.cmake:70]: These settings are optional and meant to be [passed to cmake by the user](https://github.com/TheLartians/ModernCppStarter#ccache). They should not be enabled by default.

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -57,15 +61,16 @@ endif()
...
-# Note: doctest and similar testing frameworks can automatically configure CMake tests For other
-# testing frameworks add the tests target instead: ADD_TEST(GreeterTests GreeterTests)
+# Note: doctest and similar testing frameworks can automatically configure CMake tests.
+# include(${doctest_SOURCE_DIR}/scripts/cmake/doctest.cmake)
+# doctest_discover_tests(GreeterTests)
 
-include(${doctest_SOURCE_DIR}/scripts/cmake/doctest.cmake)
-doctest_discover_tests(GreeterTests)
+# For other testing frameworks add the tests target instead:
+ADD_TEST(GreeterTests GreeterTests)
```
Comment (TheLartians) [inline test/CMakeLists.txt:69]: `doctest_discover_tests` is a great feature, why remove it?

**Review by TheLartians — CHANGES_REQUESTED** [review]
Hey, thanks for the PR! I see some good changes here, however we should think about what is a good for the starter to include. It should be seen as a project with examples for a few best practices, but only create as little example code as necessary to make it easier to adjust for all the different use-cases. 

> How can I handle this with CPMAddPackage()?

CPM.cmake simply adds the library's standard CMakeLists file. Unfortunately, if a library developer doesn't create a namespaced target alias there, then there won't be any imported either, so we can only use the standard target.

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.14...3.19 FATAL_ERROR)
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:1]: this adapt the policy to the currently used cmake version

I tested with 3.19

File: CMakeLists.txt (hunk 1 above)
Comment (ClausKlein) [reply CMakeLists.txt:64]: the library target may created in on `CMakeLists.txt`

and the source be added later in a sub project added with `add_subdirectory()`

File: cmake/tools.cmake (hunk 4 above)
Comment (ClausKlein) [reply cmake/tools.cmake:70]: IMHO no,  use cache and ninja to get the fasted roundtrip time

File: CMakeLists.txt [hunk 7]
```diff
@@ -30,27 +36,38 @@ CPMAddPackage(
...
 # ---- Create library ----
 
 # Note: for header-only libraries change all PUBLIC flags to INTERFACE and create an interface
-# target: add_library(Greeter INTERFACE) set_target_properties(Greeter PROPERTIES
-# INTERFACE_COMPILE_FEATURES cxx_std_17)
-
-add_library(Greeter ${headers} ${sources})
+# target:
+add_library(Greeter)
+add_library(${PROJECT_NAME}::Greeter ALIAS Greeter)
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:61]: > Hey, thanks for the PR! I see some good changes here, however we should think about what is a good for the starter to include. It should be seen as a project with examples for a few best practices, but only create as little example code as necessary to make it easier to adjust for all the different use-cases.
> 
> > How can I handle this with CPMAddPackage()?
> 
> CPM.cmake simply adds the library's standard CMakeLists file. Unfortunately, if a library developer doesn't create a namespaced target alias there, then there won't be any imported either, so we can only use the standard target. To fix this, you could ask the library developer to add an alias target to his CMakeLists.

from `/usr/local/lib/cmake/fmt/fmt-targets.cmake`

```cmake
# Create imported target fmt::fmt
add_library(fmt::fmt SHARED IMPORTED)

set_target_properties(fmt::fmt PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "FMT_LOCALE;FMT_SHARED"
  INTERFACE_COMPILE_FEATURES "cxx_variadic_templates"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
)

# Create imported target fmt::fmt-header-only
add_library(fmt::fmt-header-only INTERFACE IMPORTED)

set_target_properties(fmt::fmt-header-only PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "FMT_HEADER_ONLY=1"
  INTERFACE_COMPILE_FEATURES "cxx_variadic_templates"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
)

``

File: CMakeLists.txt (hunk 2 above)
Comment (ClausKlein) [reply CMakeLists.txt:44]: I understand, but IMHO the template is to simple without a link dependency

File: CMakeLists.txt [hunk 8]
```diff
@@ -18,6 +18,12 @@ if(PROJECT_SOURCE_DIR STREQUAL PROJECT_BINARY_DIR)
   )
 endif()
 
+# ---- Project settings ----
+
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS NO)
+
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:26]: The minimum std for the hole project, but I will make it optional

File: CMakeLists.txt [hunk 9]
```diff
@@ -30,27 +36,38 @@ CPMAddPackage(
...
+find_package(fmt)
+if(NOT TARGET fmt::fmt)
+  # FIXME this add a target without namespace! CK
+  CPMAddPackage(
+    NAME fmt
+    GIT_TAG 7.1.3
+    GITHUB_REPOSITORY fmtlib/fmt
+  )
+endif()
+
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:48]: this is the main point, it does not work!

File: test/CMakeLists.txt (hunk 5 above)
Comment (ClausKlein) [reply test/CMakeLists.txt:69]: If doctest is installed on build host and used, this does not work!

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:63]: now, if get a new problem:

```shell
CMake Error at CMakeLists.txt:45 (add_executable):
   Target "GreeterTests" links to target "fmt::fmt-header-only" but the target
   was not found.  Perhaps a find_package() call is missing for an IMPORTED
   target, or an ALIAS target is missing?

 Generating done
```

[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:63]: 2 things are wrong here:

the permissive option?

and this seems a cmake bug: the Link only export!

**Comment by ClausKlein** [comment]
Too much noise, sorry. I will clean the patch and than an other try ...

One Question: why do you use so much `set_target_properties() `if the is a clean cmake command api?

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (TheLartians) [reply CMakeLists.txt]: The permissive option comes from [a generator expression in the main CMakeLists.txt](https://github.com/TheLartians/ModernCppStarter/blob/b58e0717f4a60274e5bdbaa32054cdd738933369/CMakeLists.txt#L51). This enables permissive mode in MSVC only, which apparently helps with standard compliance. In your case the option is disabled, so it should have no influence on the library.

Hm the export could be due to an inconsistency in the exported targets and the ones declared in `fmt`'s CMakeLists file. I think we can work around it from our side, e.g. by conditionally creating an alias target. But can't say for sure without checking.

**Comment by TheLartians** [comment]
Sounds good, if possible, I'd suggest also splitting it into multiple independent PRs, e.g. one for adding the `fmt` dependency and one for changing other the CMakeLists properties to allow a more focussed discussion.

> One Question: why do you use so much set_target_properties() if there is a clean cmake command api?

And with clean api, you mean using `target_compile_features()` vs `set_target_properties`? No good reason really, it was just one of multiple ways to get the job done in CMake. Could you elaborate how you consider the `target_compile_features` cleaner?

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt]: fixed in my new PR https://github.com/TheLartians/ModernCppStarter/pull/80

**Pull request closed without merging**